### PR TITLE
Support lambdas for task actions

### LINKED ISSUE #17327: Support serializable lambdas for up-to-date checks
Gradle currently doesn't support Java lambdas (aka invoke dynamic implementations of classes) since it is unable to track the implementation for the classes the JVM generates at runtime. For serializable lambdas, Gradle could extract enough information to track the implementation and therefore support those as task actions. This is especially interesting, since Kotlin 1.5 by default generates invoke dynamic calls when a lambda implements a SAM type.

As a next step we could automatically make the lambdas we use for doLast etc. serializable.

---
cc: @gradle/execution

### PR DESCRIPTION
**Pull request by alllex**
Gradle currently doesn't support Java lambdas (aka invoke dynamic implementations of classes) since it is unable to track the implementation for the classes the JVM generates at runtime. 

This PR allows tracking the implementation of serializable lambdas to support those as task actions. This is especially interesting, since Kotlin 1.5 by default generates invoke dynamic calls when a lambda implements a SAM type.

Configuration cache decoration rewrites the bytecode of plugin classes making sure that all lambdas that implement an `Action` are serializable. For serializable lambdas we can obtain the SerializedLambda object that contains enough information to track lambda’s implementation:

https://github.com/gradle/gradle/issues/17327#issuecomment-851590456

We add a new kind of lambda snapshot to distinguish between serializable and non-serializable lambdas. We preserve the warning for the non-serializable lambdas, although it should appear in much fewer cases.

Fixes #17327

### CODE REVIEW

**Comment by alllex** [comment]
@bot-gradle test qfl

**Comment by alllex** [comment]
@bot-gradle test this

**Comment by alllex** [comment]
@bot-gradle test this

**Comment by alllex** [comment]
@bot-gradle test this

**Comment by alllex** [comment]
@bot-gradle test this

File: subprojects/snapshots/src/test/groovy/org/gradle/internal/snapshot/impl/ImplementationSnapshotTest.groovy [hunk 1]
```diff
@@ -37,15 +45,73 @@ class ImplementationSnapshotTest extends Specification {
...
+        description              | implementationFactory
+        'same serializable'      | { -> createLambdaSnapshot(SHARED_SERIALIZABLE_LAMBDA) }
+        'serializable'           | { -> createLambdaSnapshot(TestLambdas.createSerializableLambda()) }
+        'method reference'       | { -> createLambdaSnapshot(TestLambdas.createMethodRefLambda()) }
+        'static field capturing' | { -> createLambdaSnapshot(TestLambdas.createClassCapturingLambda()) }
+        'instance capturing'     | { -> createLambdaSnapshot(TestLambdas.createInstanceCapturingLambda()) }
+    }
+
+    def "implementation snapshots are not equal for lambdas of different functional interfaces"() {
+        def actionLambdaSnap = createLambdaSnapshot(TestLambdas.createMethodRefLambda())
```
Comment (wolfs) [inline subprojects/snapshots/src/test/groovy/org/gradle/internal/snapshot/impl/ImplementationSnapshotTest.groovy:100]:
```suggestion
        def actionLambdaSnapshot = createLambdaSnapshot(TestLambdas.createMethodRefLambda())
```

I'd write out `snapshot`, there isn't much benefit in using an abbreviation.

File: subprojects/snapshots/src/test/groovy/org/gradle/internal/snapshot/impl/ImplementationSnapshotTest.groovy [hunk 2]
```diff
@@ -37,15 +45,73 @@ class ImplementationSnapshotTest extends Specification {
...
+        'unknown classloader for untrackable lambda' | { -> ImplementationSnapshot.of('SomeClass$$Lambda$3/23501234324', SHARED_UNTRACKABLE_LAMBDA, null) }
+    }
+
+    def "cannot hash unknown implementation snapshot"() {
+        def snapshot = ImplementationSnapshot.of("SomeClass", null)
+        when:
+        snapshot.appendToHasher(Mock(Hasher))
+
+        then:
+        thrown(UnsupportedOperationException)
```
[the remark this answers is not in the document]
Comment (wolfs) [reply subprojects/snapshots/src/test/groovy/org/gradle/internal/snapshot/impl/ImplementationSnapshotTest.groovy:67]: https://spockframework.org/spock/docs/2.0/interaction_based_testing.html#_interactions

**Comment by alllex** [comment]
@bot-gradle test qfl

**Comment by alllex** [comment]
@bot-gradle test prf pt

**Review by wolfs — APPROVED** [review]
LGTM!

**Comment by alllex** [comment]
@bot-gradle test and merge

File: subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/LambdaImplementationSnapshot.java [hunk 3]
```diff
@@ -20,60 +20,118 @@
...
+        this.implClass = implClass;
+        this.implMethodName = implMethodName;
+        this.implMethodSignature = implMethodSignature;
+        this.implMethodKind = implMethodKind;
     }
 
     @Override
     public void appendToHasher(Hasher hasher) {
-        throw new RuntimeException("Cannot hash implementation of lambda " + getTypeName());
+        hasher.putString(ImplementationSnapshot.class.getName());
```
Comment (wolfs) [inline subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/LambdaImplementationSnapshot.java:68]:
```suggestion
        hasher.putString(LambdaImplementationSnapshot.class.getName());
```

Similar to `ClassImplementationSnapshot`, we should probably use the concrete class here. Though given that the number of things we hash are different, we are already sure that we don't get a collision.

**Pull request merged**